packages/flutter/lib/src/widgets/scroll_position.dart

```diff
@@ -195,15 +197,54 @@ abstract class ScrollPosition extends ViewportOffset with ScrollMetrics {
   /// To cause the position to jump or animate to a new value, consider [jumpTo]
   /// or [animateTo].
   ///
-  /// This should not be called during layout. Consider [correctPixels] if you
-  /// find you need to adjust the position during layout.
+  /// This should not be called during layout (e.g. when setting the initial
+  /// scroll offset). Consider [correctPixels] if you find you need to adjust
+  /// the position during layout.
   @protected
   void forcePixels(double value) {
-    assert(_pixels != null);
+    assert(pixels != null);
     _pixels = value;
     notifyListeners();
   }
 
+  /// Called whenever scrolling ends, to store the current scroll offset in a
+  /// storage mechanism with a lifetime that matches the app's lifetime.
+  ///
+  /// The stored value will be used by [restoreScrollOffset] when the
+  /// [ScrollPosition] is recreated, in the case of the [Scrollable] being
+  /// disposed then recreated in the same session. This might happen, for
+  /// instance, if a [ListView] is on one of the pages inside a [TabBarView],
+  /// and that page is displayed, then hidden, then displayed again.
+  ///
+  /// The default implementation writes the [pixels] using the nearest
+  /// [PageStorage] found from the [context]'s [ScrollContext.storageContext]
+  /// property.
+  @protected
+  void saveScrollOffset() {
+    PageStorage.of(context.storageContext)?.writeState(context.storageContext, pixels);
+  }
+
+  /// Called whenever the [ScrollPosition] is created, to restore the scroll
+  /// offset if possible.
+  ///
+  /// The value is stored by [saveScrollOffset] when the scroll position
+  /// changes, so that it can be restored in the case of the [Scrollable] being
+  /// disposed then recreated in the same session. This might happen, for
+  /// instance, if a [ListView] is on one of the pages inside a [TabBarView],
+  /// and that page is displayed, then hidden, then displayed again.
+  ///
+  /// The default implementation reads the value from the nearest [PageStorage]
+  /// found from the [context]'s [ScrollContext.storageContext] property, and
+  /// sets it using [correctPixels], if [pixels] is still null.
+  @protected
+  void restoreScrollOffset() {
+    if (pixels == null) {
```

Comment:
Please test that the restored scroll offset wins over the initial scroll offset.